server/permalinks.go

```diff
@@ -0,0 +1,169 @@
+package main
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/xanzy/go-gitlab"
+)
+
+// maxPermalinkReplacements sets the maximum limit to the number of
+// permalink replacements that can be performed on a single message.
+const maxPermalinkReplacements = 10
+
+const permalinkReqTimeout = 5 * time.Second
+
+// maxPreviewLines sets the maximum number of preview lines that will be shown
+// while replacing a permalink.
+const maxPreviewLines = 10
+
+// permalinkLineContext shows the number of lines before and after to show
+// if the link points to a single line.
+const permalinkLineContext = 3
+
+// replacement holds necessary info to replace GitLab permalinks
+// in messages with a code preview block.
+type replacement struct {
+	index         int    // Index of the permalink in the string
+	word          string // The permalink
+	permalinkData permalinkInfo
+}
+
+type permalinkInfo struct { // Holds the necessary metadata of a permalink
+	haswww string
+	commit string
+	user   string
+	repo   string
+	path   string
+	line   string
+}
+
+// getPermalinkReplacements returns the permalink replacements that need to be performed
+// on a message. The returned slice is sorted by the index in ascending order.
+func (p *Plugin) getPermalinkReplacements(msg string) []replacement {
+	// Find the permalinks from the msg using a regex
+	matches := gitlabPermalinkRegex.FindAllStringSubmatch(msg, -1)
+	indices := gitlabPermalinkRegex.FindAllStringIndex(msg, -1)
+	var replacements []replacement
+	for i, m := range matches {
+		// Have a limit on the number of replacements to do
+		if i > maxPermalinkReplacements {
+			break
+		}
+		word := m[0]
+		index := indices[i][0]
+		r := replacement{
+			index: index,
+			word:  word,
+		}
+		// Ignore if the word is inside a link
+		if isInsideLink(msg, index) {
+			continue
+		}
+		// Populate the permalinkInfo with the extracted groups of the regex
+		for j, name := range gitlabPermalinkRegex.SubexpNames() {
+			if j == 0 {
+				continue
+			}
+			switch name {
+			case "haswww":
+				r.permalinkData.haswww = m[j]
+			case "user":
+				r.permalinkData.user = m[j]
+			case "repo":
+				r.permalinkData.repo = m[j]
+			case "commit":
+				r.permalinkData.commit = m[j]
+			case "path":
+				r.permalinkData.path = m[j]
+			case "line":
+				r.permalinkData.line = m[j]
+			}
+		}
+		replacements = append(replacements, r)
+	}
+	return replacements
+}
+
+func (p *Plugin) processReplacement(r replacement, glClient *gitlab.Client, wg *sync.WaitGroup, markdownForPermalink []string, index int) {
```

Comment:
0/5 maybe worth a quick comment that concurrently modifying unique slice elements is not racy.

Reply:
Done